doc/developer/design/2024_10_08_expression_cache.md

```diff
@@ -0,0 +1,135 @@
+    /// will not change in-between restarts as result of DDL, as long as `global_id` exists.
+    fn get_deployed_expression(&self, global_id: GlobalId, expression_type: ExpressionType) -> Option<Bytes>;
+
+    /// Returns the `expression_type` of `global_id` based on the current catalog contents of
+    /// `deploy_generation`. This may change in-between restarts as result of DDL.
+    fn get_durable_expression(&self, deploy_generation: u64, global_id: GlobalId, expression_type: ExpressionType) -> Option<Bytes>;
+
+    /// Durably inserts `expression`, with key `(deploy_generation, global_id, expression_type)`.
+    ///
+    /// Panics if `(deploy_generation, global_id, expression_type)` already exists.
+    fn insert_expression(&mut self, deploy_generation: u64, global_id: GlobalId, expression_type: ExpressionType, expression: Bytes);
+
+    /// Durably remove and return all entries in `deploy_generation` that depend on an ID in
+    /// `dropped_ids`.
+    fn invalidate_entries(&mut self, deploy_generation: u64, dropped_ids: BTreeSet<GlobalId>) -> Vec<(GlobalId, ExpressionType)>;
+    
+    /// Durably removes all entries in `deploy_generation`.
+    fn remove_deploy_generation(&mut self, deploy_generation: u64);
+    
+    /// Remove all entries that depend on a global ID that is not present in `txn`.
+    fn reconcile(&mut self, txn: mz_catalog::durable::Transaction);
+}
+```
+
+### Startup
+
+Below is a detailed set of steps that will happen in startup.
+
+1. Call `ExpressionCache::reconcile` to remove any invalid entries.
+2. While opening the catalog, for each object:
+    a. If the object is present in the cache, read the cached optimized expression via
+       `ExpressionCache::get_durable_expression`.
+    b. Else generate the optimized expressions and insert the expressions via
+       `ExpressionCache::insert_expression`.
+3. If in read-write mode, call `ExpressionCache::remove_deploy_generation` to remove the previous 
+   deploy generation.
+
+### DDL - Create
+
+1. Execute catalog transaction.
+2. Update cache via `ExpressionCache::insert_expression`. 
+
+### DDL - Drop
+1. Execute catalog transaction.
+2. Invalidate cache entries via `ExpressionCache::invalidate_entries`.
+3. Re-compute and repopulate cache entries that depended on dropped entries via 
+   `ExpressionCache::insert_expression`.
+
+### File System Implementation
+
+One potential implementation is via the filesystem of an attached durable storage to environmentd.
+Each cache entry would be saved as a file of the format
+`/path/to/cache/<deploy_generation>/<global_id>/<expression_type>`.
+
+#### Pros
+- No need to worry about coordination across K8s pods.
+- Bulk deletion is a simple directory delete.
+
+#### Cons
+- Need to worry about Flushing/fsync.
+- Need to worry about concurrency.
+- Need to worry about atomicity.
+- Need to worry about mocking things in memory for tests.
+- If we lose the pod, then we also lose the cache.
+
+### Persist implementation
```

Comment:
A second idea that Aljsocha and I had in our 1:1 today is to use persist, but not with the normal consensus and blob impls (e.g. use FileBlob, ??? on the Consensus impl). That could potentially get you a persist impl where you don't need to worry about coordination (e.g. if they're both pointed at local fs)